src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,1050 @@
+        // No need to do anything for an instruction that has been squashed.
+        return;
+    }
+
+    DPRINTF(SDCPUThreadEvent, "attemptFetch()\n");
+
+    // TODO check for interrupts
+    // NOTE: You cannot take an interrupt in the middle of a macro-op.
+    //      (How would you recover the architectural state since you are in a
+    //       non-architectural state during a macro-op?)
+    //      Thus, we must check instruction->isDelayedCommit() and make sure
+    //      that interrupts only happen after an instruction *without* the
+    //      isDelayedCommit flag is committed.
+    // TODO something something PCEventQueue?
+
+    TheISA::PCState pc_value = inst_ptr->pcState();
+
+    // Retrieve the value of the speculative PC as a raw Addr value, aligned
+    // to MachInst size, but offset by the amount of data we've fetched thus
+    // far, since multiple fetches may be needed to feed one decode.
+    const Addr pc_addr = pc_value.instAddr();
+    const Addr req_vaddr = (pc_addr & BaseCPU::PCMask) + fetchOffset;
+
+    // If HIT in fetch buffer, we can skip sending a new request, and decode
+    // from the buffer immediately.
+    if (fetchBufBase
+     && req_vaddr >= fetchBufBase
+     && req_vaddr + sizeof(TheISA::MachInst)
+        <= fetchBufBase + fetchBuf.size()) {
+        DPRINTF(SDCPUThreadEvent, "Grabbing fetch data for vaddr(%#x) from "
+                                  "buffer.\n", req_vaddr);
+
+        const TheISA::MachInst inst_data =
+            *reinterpret_cast<TheISA::MachInst*>(fetchBuf.data()
+                                               + (req_vaddr - fetchBufBase));
+
+        onInstDataFetched(inst, inst_data);
+        return;
+    } // Else we need to send a request to update the buffer.
+
+    DPRINTF(SDCPUThreadEvent, "Preparing inst translation request at %#x\n",
+                              req_vaddr);
+
+    const RequestPtr fetch_req = std::make_shared<Request>();
+
+    // Set Ids for tracing requests through the system
+    fetch_req->taskId(_cpuPtr->taskId());
+    fetch_req->setContext(_committedState->contextId()); // May need to store
+                                                         // speculative variant
+                                                         // if we allow out-of-
+                                                         // order execution
+                                                         // between context
+                                                         // changes?
+
+    const Addr new_buf_base = req_vaddr & fetchBufMask;
+    const unsigned req_size = fetchBuf.size();
+
+    fetch_req->setVirt(0, new_buf_base, req_size,
+                       Request::INST_FETCH, _cpuPtr->instMasterId(), pc_addr);
+
+    auto callback = [this, inst](Fault f, const RequestPtr& r) {
+        onPCTranslated(inst, f, r);
+    };
+
+    if (!_cpuPtr->requestInstAddrTranslation(fetch_req, getThreadContext(),
```

Comment:
It's a little weird that you don't translate things that hit in the fetch buffer. It could be a problem for two reasons:
1) In some sense you're "speculating" that it's a TLB hit. It might not be. If you are going to skip the translation, then you need a mechanism to *guarantee* that it's a TLB hit.
2) From a software point of view, you're sometimes bypassing the TLB and sometimes not. That's also a little weird. It makes the software logic more complicate (e.g., if you try to explain it with a diagram).